python_modules/dagster/dagster/_core/definitions/parent_asset_graph_differ.py

```diff
@@ -0,0 +1,144 @@
+    _branch_asset_graph_load_fn: Optional[Callable[[], "ExternalAssetGraph"]]
+    _parent_asset_graph: Optional["ExternalAssetGraph"]
+    _parent_asset_graph_load_fn: Optional[Callable[[], "ExternalAssetGraph"]]
+
+    def __init__(
+        self,
+        branch_asset_graph: Union["ExternalAssetGraph", Callable[[], "ExternalAssetGraph"]],
+        parent_asset_graph: Optional[
+            Union["ExternalAssetGraph", Callable[[], "ExternalAssetGraph"]]
+        ] = None,
+    ):
+        if parent_asset_graph is None:
+            # if parent_asset_graph is None, then the asset graph in the branch deployment does not exist
+            # in the parent deployment
+            self._parent_asset_graph = None
+            self._parent_asset_graph_load_fn = None
+        elif isinstance(parent_asset_graph, ExternalAssetGraph):
+            self._parent_asset_graph = parent_asset_graph
+            self._parent_asset_graph_load_fn = None
+        else:
+            self._parent_asset_graph = None
+            self._parent_asset_graph_load_fn = parent_asset_graph
+
+        if isinstance(branch_asset_graph, ExternalAssetGraph):
+            self._branch_asset_graph = branch_asset_graph
+            self._branch_asset_graph_load_fn = None
+        else:
+            self._branch_asset_graph = None
+            self._branch_asset_graph_load_fn = branch_asset_graph
+
+    @classmethod
+    def from_external_repositories(
+        cls,
+        code_location_name: str,
+        repository_name: str,
+        branch_workspace: BaseWorkspaceRequestContext,
+        parent_workspace: Optional[BaseWorkspaceRequestContext] = None,
+    ) -> Optional["ParentAssetGraphDiffer"]:
+        """Constructs a ParentAssetGraphDiffer for a particular repository in a code location for two
+        deployment workspaces, the parent deployment and the branch deployment. If the
+        parent_workspace is None, then we are not in a branch deployment and will not create a ParentAssetGraphDiffer.
+        """
+        if parent_workspace is not None:
+            branch_repo = _get_external_repo_from_context(
+                branch_workspace, code_location_name, repository_name
+            )
+            if branch_repo is None:
+                raise DagsterInvariantViolationError(
+                    f"Repository {repository_name} does not exist in code location {code_location_name} for the branch deployment."
+                )
+            parent_repo = _get_external_repo_from_context(
+                parent_workspace, code_location_name, repository_name
+            )
+            return ParentAssetGraphDiffer(
+                branch_asset_graph=lambda: ExternalAssetGraph.from_external_repository(branch_repo),
+                parent_asset_graph=(
+                    lambda: ExternalAssetGraph.from_external_repository(parent_repo)
+                )
+                if parent_repo is not None
+                else None,
```

Comment:
I think we might want to just use `ExternalAssetGraph.from_workspace` and remove this code location scoping, probably call this method `from_workspaces`

what led you to code location scope?

Reply:
I found that if i just did from workspaces, I ran into issues with duplicate asset keys. I think it might be a larger issue in the `AssetGraph` maybe?

Let's say the parent deployment looks like this:
```
code_location_1: 
     - asset_1
     - asset_2
     - asset3
code_location_2:
    - asset_1
    - asset_2
    - asset_b
```

and the branch deployment we only modify `asset_1` and `asset_2` in `code_location_1`.

I found that when i did `from_workspace` the resulting AssetGraph would only have one occurrence of `asset_1` and `asset_2`, and that it could be from `code_location_2` instead of `code_location_1` which would make the comparison of the changes incorrect